### PR TITLE
fix: etcd compact err【do not merge me! only for discussion】

### PR DESCRIPTION
**Pull request by Revolyssup**
### Description

Fixes # (issue)

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/core/config_etcd.lua [hunk 1]
```diff
@@ -328,21 +327,37 @@ local function sync_data(self)
...
+    local rev
+    if res and res.body and res.body.header and res.body.header.revision then
+        rev = tonumber(res.body.header.revision)
+    end
     local dir_res, err = waitdir(self.etcd_cli, self.key, self.prev_index + 1, self.timeout)
     log.info("waitdir key: ", self.key, " prev_index: ", self.prev_index + 1)
     log.info("res: ", json.delay_encode(dir_res, true))
-
     if not dir_res then
-        if err == "compacted" then
```
Comment (nic-chen) [inline apisix/core/config_etcd.lua:336]: please don't remove these codes, we should keep them.

Comment (Revolyssup) [reply apisix/core/config_etcd.lua:336]: okay

File: apisix/core/config_etcd.lua [hunk 2]
```diff
@@ -328,21 +327,43 @@ local function sync_data(self)
...
     if not dir_res then
+        if err == "timeout" then
+            log.info("updating prev_index from "..self.prev_index.." to "..rev )
+            if rev then
+                if rev == self.prev_index then
+                    goto fail
+                end
+                self.prev_index = rev
+            end
+            goto waitdir
```
Comment (nic-chen) [inline apisix/core/config_etcd.lua:355]: how about the change?

```suggestion
            log.info("updating prev_index from "..self.prev_index.." to "..rev )
            self:upgrade_version(rev)
```

File: apisix/core/config_etcd.lua [hunk 3]
```diff
@@ -328,21 +327,43 @@ local function sync_data(self)
...
+        end
         if err == "compacted" then
             self.need_reload = true
             log.warn("waitdir [", self.key, "] err: ", err,
                      ", will read the configuration again via readdir")
-            return false
+            goto fail
```
Comment (nic-chen) [inline apisix/core/config_etcd.lua:361]: we don't need to change the workflow

**Pull request closed without merging**